### ISSUE TITLE
Sitemap not outputting

### ISSUE THREAD

**Issue by k1sul1** [sybrew/the-seo-framework#451]
Apologies if this is already possible, but I wasn't able to find a filter for this, even after doing some digging in the code.

As the title states, I'd like to change the url of <loc> entries in the sitemap. Why? Because I'm running WordPress on wp.mydomain.com, and the frontend on mydomain.com. This means that the sitemap links point to wp.mydomain.com, when I need to have them point to mydomain.com.

This is "somewhat" common in single page applications (n=2 :trollface:). Google seems to be fine with the sitemap redirecting to another subdomain. 

Is there a way to do this now, that doesn't involve using ob_buffer hacks? I tried messing with WP_SITEURL constant without luck.

**Comment by sybrew** [comment]
Hello!

`WP_HOME` defines the homepage URL.
`WP_SITEURL` defines the WordPress root file location.

Both can be set manually via the general settings of WordPress, via "WordPress Address (URL)" and "Site Address (URL)" respectively.

Now, I'm a bit confused: We use `WP_HOME` to generate the URLs; so, they shouldn't point to your WordPress installation folder. We also use the WordPress rewrite engine to output the sitemap (until the next version), so the sitemap should work only on the root-domain, not the subdomain.

Are you finding difficulties outputting your sitemap at `example.com/sitemap.xml`, but it is outputted "correctly" at `wp.example.com/sitemap.xml`? And, are our URLs pointing to the subdomain, instead of the root-domain?

If either is the case, then I believe you touched a dial too many 😅, breaking the rewrite engine.

Please refer to [Giving WordPress Its Own Directory](https://wordpress.org/support/article/giving-wordpress-its-own-directory/) and retrace your steps.

**Comment by k1sul1** [comment]
You understood wrong :slightly_smiling_face: I haven't touched a single dial. I don't think the edited title describes the situation at all. There's nothing wrong with the sitemap output, I just want to change it a bit. In another project that we did, we had to use Yoast, and we used `wpseo_sitemap_url` filter to get this done.  

I've basically disabled WordPress frontend by making a very barebones theme, that redirects to my single page application. WordPress is totally unaware of this, as I want it to be. It thinks that it's being used from wp.mydomain.com, which is partially true, as the backend is used through that URL.

All content is viewed on mydomain.com. I used Nginx to get the sitemap available at mydomain.com/sitemap.xml, since I can't submit a sitemap from **wp.mydomain.com** for **mydomain.com**. I can submit the sitemap, but instead of containing links to my single page application (that is rendered on the server for bots), it contains links to my WP backend.

**Comment by sybrew** [comment]
Oh, my bad! ☹️ 

I still think we're both confusing some things.

First, please note that we cache the sitemap via transient caching. When you change stuff in the code of your site, we may not be able to anticipate a flushing action. So, it's best to re-save the SEO Settings, which will force such a flush. See if this updates the URLs in your sitemap.

Now, a filter much like `wpseo_sitemap_url` will be shipped in the next update. See https://github.com/sybrew/the-seo-framework/issues/426#issuecomment-503624483, albeit, we only cover endpoints, not fully qualified URIs.

However, neither our solution nor the one you pointed from Yoast SEO change the URLs inside the sitemap. This is where my confusion still stands, and it's why I changed the title of this issue (to evidently something unrelated).

To change the URLs inside the sitemap, you'd want to filter the WordPress default URL handler. This all goes via function [get_home_url()](https://developer.wordpress.org/reference/functions/get_home_url/).

That function uses the option `home_url`, which is equal to (whichever comes first):
1. The constant `WP_HOME`, default undefined. You can set this using `define( 'WP_HOME', 'https://example.com' );`.
2. The "WordPress Address (URL)" input at `/wp-admin/options-general.php`.

Change either of those, and the URLs should point towards what you put in.
However, if changing those mess up your site, then you may want to filter the home URL:
```php
// Force changing wp.example.com to example.com. You may find preg_replace() to do this more safely, but I don't see a case for that here.
add_filter( 'home_url', function( $url ) {
	return str_ireplace( 'https://wp.example.com`, `https://example.com`, $url );
}, PHP_INT_MAX, 1 );
```

Anecdotally, running that filter will slow down your site a bit; it runs pretty much everywhere.

If I still have the wrong impression on your inquiry, please send me over your domain name, so I can take in what's going on. You may share it confidentially via [my contact page](https://theseoframework.com/contact/). Cheers!

**Comment by k1sul1** [comment]
I had WP_HOME and WP_SITEURL mixed, not anymore. 

I guess I could set WP_HOME to mydomain.com when the request url matches for a sitemap request. Setting it pretty much anywhere else will break everything :slightly_smiling_face:  

I see one possible problem in that approach, and you mentioned it already; you cache sitemaps. If the sitemaps are always created "on-demand" and then cached, instead of generating them every time I click save, then it will probably work fine. 

I'll give it a try, but I don't think I've time to look into that site before next week. I just checked the site that uses Yoast, and the links point to the right place. AFAIK `wpseo_sitemap_url` is the only thing that was done there. I wrote the backend code myself, but I'm not associated with that project anymore.

Thanks for your help! :slightly_smiling_face:
